backend-shared/config/db/util/utils_test.go

```diff
@@ -1,12 +1,881 @@
+			ctx, dbQueries, log, workSpaceUid, err = initialSetUp()
+			Expect(err).To(BeNil())
+
+			// ----------------------------------------------------------------------------
+			By("Create DeploymentToApplicationMapping.")
+			// ----------------------------------------------------------------------------
+
+			kubernetesToDBResourceMapping = db.KubernetesToDBResourceMapping{
+				KubernetesResourceType: db.K8sToDBMapping_Namespace,
+				KubernetesResourceUID:  string(workSpaceUid),
+				DBRelationType:         db.K8sToDBMapping_GitopsEngineCluster,
+				DBRelationKey:          "dummy-relation-key",
+			}
+
+			err = dbQueries.CreateKubernetesResourceToDBResourceMapping(ctx, &kubernetesToDBResourceMapping)
+			Expect(err).To(BeNil())
+
+			// ----------------------------------------------------------------------------
+			By("No GitopsEngineCluster instance should be returned.")
+			// ----------------------------------------------------------------------------
+
+			gitopsEngineCluster, err = GetGitopsEngineClusterByKubeSystemNamespaceUID(ctx, string(workSpaceUid), dbQueries, log)
+
+			Expect(err).To(BeNil())
+			Expect(gitopsEngineCluster).To(BeNil())
+		})
+
+		It("Should return GitopsEngineCluster, if KubernetesToDBResourceMapping and GitopsEngineCluster exist in DB.", func() {
+
+			ctx, dbQueries, log, workSpaceUid, err = initialSetUp()
+			Expect(err).To(BeNil())
+
+			// ----------------------------------------------------------------------------
+			By("First create new DeploymentToApplicationMapping and GitopsEngineCluster to be used later.")
+			// ----------------------------------------------------------------------------
+
+			gitopsEngineCluster, _, err = GetOrCreateGitopsEngineClusterByKubeSystemNamespaceUID(ctx, string(workSpaceUid), dbQueries, log)
+
+			Expect(err).To(BeNil())
+
+			// ----------------------------------------------------------------------------
+			By("Existing GitopsEngineCluster instance should be returned.")
+			// ----------------------------------------------------------------------------
+
+			retriveGitopsEngineCluster, err := GetGitopsEngineClusterByKubeSystemNamespaceUID(ctx, string(workSpaceUid), dbQueries, log)
+
+			Expect(err).To(BeNil())
+			Expect(gitopsEngineCluster).To(Equal(retriveGitopsEngineCluster))
+
+			// ----------------------------------------------------------------------------
+			By("To cleanup the resource created by test")
+			// ----------------------------------------------------------------------------
+
+			// Fetch KubernetesToDBResourceMapping to be cleaned in AfterEach.
+			kubernetesToDBResourceMapping, _, err = findKubernetesToDBResourceMappingInTable(ctx, dbQueries, gitopsEngineCluster.Gitopsenginecluster_id, db.K8sToDBMapping_GitopsEngineCluster)
+			Expect(err).To(BeNil())
+		})
+	})
+
+	Context("Testing for GetOrCreateGitopsEngineInstanceByInstanceNamespaceUID function.", func() {
```

Comment:
It can be tough to tell whether the story criteria items are being met when it comes to writing tests, due to the nature of test writing.

In order to tell if a particular part of a test is being covered, you can verify this by running `make test`, which will produce a coverage.out file. Then, in VS Code, select `View` > `Command Palette`, then `Go: Apply Cover Profile`, and specify the absolute path of the coverage.out file.

I did that, and it appears that the following story criteria items are not being fully tested:
- To test: if a KubernetesToDBResourceMapping mapping exists, and the engine instance row doesn't exist: verify that the function deletes the old mapping, and verify it creates a new KubernetesToDBResourceMapping and GitOpsEngineInstance row
- To test: if a KubernetesToDBResourceMapping mapping exists, and the engine instance row exists: verify that the function does not create a new KubernetesToDBResourceMapping or a new GitOpsEngineInstanceRow... it; it should just retrieve existing values from database

Then, If you follow the above steps, then you open `GetOrCreateGitopsEngineInstanceByInstanceNamespaceUID` function, you will see that 'Scenario C' and 'Scenario D' code blocks are not covered. This corresponds to the two test items above. (Don't worry about 'Scenario B'.)

(https://issues.redhat.com/browse/GITOPSRVCE-135)

Reply:
Actually these story criteria items are covering these tests. 
`To test: if a KubernetesToDBResourceMapping mapping exists, and the engine instance row doesn't exist: verify that the function deletes the old mapping, and verify it creates a new KubernetesToDBResourceMapping and GitOpsEngineInstance row`
**It("Should delete existing KubernetesToDBResourceMapping and create new KubernetesToDBResourceMapping and GitopsEngineInstance.", func(){}**

`To test: if a KubernetesToDBResourceMapping mapping exists, and the engine instance row exists: verify that the function does not create a new KubernetesToDBResourceMapping or a new GitOpsEngineInstanceRow... it; it should just retrieve existing values from database`
**It("Should return existing KubernetesToDBResourceMapping and GitopsEngineInstance, instead of creating new.", func(){}**

In test coverage report `Scenario C` is not covered, but what I understand, in current implementation this code block will not execute. 
This block will execute if instance exists, but there is no mapping for it. In that case condition written above will execute and control will not go upto block written for Section C

```
// <This condition will satisfy because mapping is not available, 
// hence it will not go in else block and instance will not be fetched from DB, 
// it means instance will be Nil>

if err := dbq.GetDBResourceMappingForKubernetesResource(ctx, dbResourceMapping); err != nil {
	dbResourceMapping = nil
} else {
	// This will search for GitOpsEngineCluster and GitOpsEngineInstance
}

// <Mapping and instance both are Nil so Scenario A will execute instead of Scenario C>

if dbResourceMapping == nil && gitopsEngineInstance == nil {
		// Scenario A) neither exists: create both
}  else if dbResourceMapping == nil && gitopsEngineInstance != nil {
		// Scenario C) this will happen if the instance exists, but there is no mapping for it
}
```